### ISSUE TITLE
Getting:- INVALID_CLIENT: Invalid redirect URI

### ISSUE THREAD

**Issue by Daftjoe** [malcolmsailor/spotify_filter_played#1]
I'm very knew to coding and quite possibly doing something wrong. 
After entering my client and secret id's the script fails as follows. The browser opens and gives me INVALID_CLIENT: Invalid redirect URI
in the body of the page.

The address bar contains:- (I've hidden my client_id)
https://accounts.spotify.com/authorize?client_id=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx&redirect_uri=https%3A%2F%2Fexample.com%2Fcallback&response_type=code&scope=playlist-modify-private+playlist-modify-public+playlist-read-collaborative+playlist-read-private+ugc-image-upload+user-follow-modify+user-follow-read+user-library-modify+user-library-read+user-modify-playback-state+user-read-currently-playing+user-read-email+user-read-playback-position+user-read-playback-state+user-read-private+user-read-recently-played+user-top-read&state=NK4K2PpfNUgReoIAbwnAYntz2M_PUSCOylcGSCQPnp4&show_dialog=true

Pasting this into the script gives errors

Traceback (most recent call last):
  File "c:\spotify_filter_played-main\spotify_filter_played.py", line 456, in <module>
    main()
  File "c:\spotify_filter_played-main\spotify_filter_played-main\spotify_filter_played.py", line 422, in main
    client_id, client_secret, user_token = init_auth(args.reauthenticate)
  File "c:\spotify_filter_played-main\spotify_filter_played-main\spotify_filter_played.py", line 106, in init_auth
    user_token = tk.prompt_for_user_token(
  File "c:\Users\xxx\AppData\Local\Programs\Python\Python310\lib\site-packages\tekore\_auth\util.py", line 245, in prompt_for_user_token
    return auth.request_token(url=redirected)
  File "c:\Users\xxx\AppData\Local\Programs\Python\Python310\lib\site-packages\tekore\_auth\util.py", line 166, in request_token
    code = parse_code_from_url(url)
  File "c:\Users\xxx\AppData\Local\Programs\Python\Python310\lib\site-packages\tekore\_auth\util.py", line 49, in parse_code_from_url
    return _parse_url_param(url, "code")
  File "c:\Users\xxx\AppData\Local\Programs\Python\Python310\lib\site-packages\tekore\_auth\util.py", line 27, in _parse_url_param
    raise KeyError(msg)
KeyError: 'Passed URL contains no parameter `code`!'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\spotify_filter_played-main\spotify_filter_played-main\spotify_filter_played.py", line 462, in <module>
    assert exc_type.__module__ in (
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\spotify_filter_played-main\spotify_filter_played-main\spotify_filter_played.py", line 474, in <module>
    assert any("no attribute 'CLOSED'" in arg for arg in exc_value.args)
AssertionError

Any help you can give would be appreciated
Hopefully it's not me being an idiot
Thanks
Eric

**Comment by malcolmsailor** [comment]
Hi, thanks for your issue!

I use `tekore` to handle the Spotify requests, so let's see if their authorization examples from [https://tekore.readthedocs.io/en/stable/getting_started.html](https://tekore.readthedocs.io/en/stable/getting_started.html) can work for you. Are you able to run the following code snippets?

# 1

```
import tekore as tk

client_id = 'your_id_here'
client_secret = 'your_secret_here'

app_token = tk.request_client_token(client_id, client_secret)
```

# 2

```
import tekore as tk

client_id = 'your_id_here'
client_secret = 'your_secret_here'

redirect_uri = 'https://example.com/callback'

user_token = tk.prompt_for_user_token(
    client_id,
    client_secret,
    redirect_uri,
    scope=tk.scope.every
)
```

**Comment by Daftjoe** [comment]
Hi. Thanks for the suggestions.

The results are as follows:-

**1.**
Works fine and returns a token as expected.

BQCesg097hF_nGZTwELHHAcrsqWRlh5r4it4zp2CxBxZM3jyJbcBlJs8HIcT02GQ6AkB3bfEegTgVmipOwASwozo0pZaFy8q1FqOlKiHM-4LbpjJYkQ

**2.**
Fails as the other script does with an INVALID_CLIENT: Invalid redirect URI

Address bar contents:-

https://accounts.spotify.com/authorize?client_id=b753d36cb3d4452b9f3c6d90541cfc6a&redirect_uri=https%3A%2F%2Fexample.com%2Fcallback&response_type=code&scope=playlist-modify-private+playlist-modify-public+playlist-read-collaborative+playlist-read-private+ugc-image-upload+user-follow-modify+user-follow-read+user-library-modify+user-library-read+user-modify-playback-state+user-read-currently-playing+user-read-email+user-read-playback-position+user-read-playback-state+user-read-private+user-read-recently-played+user-top-read&state=m1NNajDMbC4dgtAyOW5M_wd8iYMkYSi284Ov8VM7v-k&show_dialog=true

Different error response from python though:-

Traceback (most recent call last):
  File "c:\spotify_filter_played-main\spotify_filter_played-main\test_stuff.py", line 18, 
in <module>
    user_token = tk.prompt_for_user_token(
  File "c:\Users\xxx\AppData\Local\Programs\Python\Python310\lib\site-packages\tekore\_auth\util.py", line 245, in prompt_for_user_token
    return auth.request_token(url=redirected)
  File "c:\Users\xxx\AppData\Local\Programs\Python\Python310\lib\site-packages\tekore\_auth\util.py", line 166, in request_token
    code = parse_code_from_url(url)
  File "c:\Users\xxx\AppData\Local\Programs\Python\Python310\lib\site-packages\tekore\_auth\util.py", line 49, in parse_code_from_url
    return _parse_url_param(url, "code")
  File "c:\Users\xxx\AppData\Local\Programs\Python\Python310\lib\site-packages\tekore\_auth\util.py", line 27, in _parse_url_param
    raise KeyError(msg)
KeyError: 'Passed URL contains no parameter `code`!'

Hope this gives an indication of what is happening

**Comment by malcolmsailor** [comment]
~If 2 is failing for you then the problem isn't with my script, but most likely with your authorization procedure or else with Tekore. You could try joining the Tekore Discord community, linked at their github [https://github.com/felix-hilden/tekore](https://github.com/felix-hilden/tekore) and asking for help debugging your issue running the code example at 2 (which comes from the Tekore docs [here](https://tekore.readthedocs.io/en/stable/getting_started.html#retrieving-a-user-token), minimally altered to include the necessary imports and authentications). You could also open a github issue on their repository but I think getting support on Discord is probably a better first port of call.~

~For now I'll close this issue, but if you are able to run the Tekore example and you still have issues with my script afterward, please feel free to come back and re-open it or a new one! (You may need to run the script with the `--reauthenticate` flag to update your id/secret afterward.) Thanks!~

Nevermind, see post about the redirect URI below.

**Comment by malcolmsailor** [comment]
Actually, I'm just now noticing the `INVALID_CLIENT: Invalid redirect URI` message. This suggests that the problem is with the `https://example.com/callback` URL that we are using. What happens if you try to visit [https://example.com/callback](https://example.com/callback) in your browser?

**Comment by malcolmsailor** [comment]
After re-reading the docs at [https://tekore.readthedocs.io/en/stable/getting_started.html#retrieving-a-user-token](https://tekore.readthedocs.io/en/stable/getting_started.html#retrieving-a-user-token) a little more carefully, it looks like you will need to add `https://example.com/callback` to the "redirect URIs" section in the application settings at [https://developer.spotify.com/dashboard/applications](https://developer.spotify.com/dashboard/applications).

Let me know if that fixes the issue for you and if so I will update the usage directions in the readme. Thanks!

**Comment by Daftjoe** [comment]
That has fixed it. I had to put https://example.com/callback in the "redirect URIs" section in the application settings. 
It now logs on and asks me for the playlist details as expected.

Thanks so much for your help.

**Comment by malcolmsailor** [comment]
Great I'm glad to hear it. If you find the script useful at all I'd be very happy to hear about it. And thank you for opening the issue! :)

**Comment by malcolmsailor** [comment]
Closed by #2
